apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryParameter.cs

```diff
@@ -0,0 +1,465 @@
+
+    /// <summary>
+    /// A parameter within the <see cref="BigqueryParameterCollection"/> of a <see cref="BigqueryCommand"/>.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// Each parameter has a name, type and value. Names are irrelevant for queries using
+    /// positional parameters. All scalar types are converted into strings
+    /// before being passed to the server, but the nature of the conversion depends on the specified type.
+    /// </para>
+    /// <para>
+    /// The following list shows the correspondence between the parameter type and valid value types.
+    /// All scalar types may be specified as a string value, which will not be validated or converted on the client,
+    /// but passed to the API as-is.
+    /// </para>
+    /// <list type="bullet">
+    ///   <item><description><c>Bool</c>: <c>System.Boolean</c></description></item>
+    ///   <item><description><c>Int64</c>: <c>System.Int16</c>, <c>System.Int32</c>, <c>System.Int64</c>, <c>System.UInt16</c>, <c>System.UInt32</c>,
+    ///   <c>System.UInt64</c></description></item>
+    ///   <item><description><c>Float64</c>: Any type valid for <c>Int64</c>, as well as <c>System.Single</c> and <c>System.Double</c></description></item>
+    ///   <item><description><c>String</c>: <c>System.String</c></description></item>
+    ///   <item><description><c>Bytes</c>: <c>System.Byte[]</c></description></item>
+    ///   <item><description><c>Date</c>: <c>System.DateTime</c>, <c>System.DateTimeOffset</c></description></item>
+    ///   <item><description><c>DateTime</c>: <c>System.DateTime</c>, <c>System.DateTimeOffset</c></description></item>
+    ///   <item><description><c>Time</c>: <c>System.DateTime</c>, <c>System.DateTimeOffset</c>, <c>System.TimeSpan</c></description></item>
+    ///   <item><description><c>Timestamp</c>: <c>System.DateTime</c>, <c>System.DateTimeOffset</c></description></item>
+    ///   <item><description><c>Array</c>: An <c>IReadOnlyList&lt;T&gt;</c> of any of the above types corresponding to the <see cref="ArrayType"/>,
+    ///   which will be inferred from the value's element type if not otherwise specified.</description></item>
+    /// </list>
+    /// <para>
+    /// If the parameter type is null, it is inferred from the value. In most cases that is obvious from the above
+    /// table, but for date and time values the process is more complicated. A <see cref="TimeSpan"/> value is
+    /// always assumed to be a <see cref="BigqueryParameterType.Time"/>, and a <see cref="DateTimeOffset"/> value is
+    /// always assumed to be a <see cref="BigqueryParameterType.Timestamp"/>. A <see cref="DateTime"/> value is assumed
+    /// to be <see cref="BigqueryParameterType.DateTime"/> if its <see cref="DateTime.Kind"/> is <see cref="DateTimeKind.Unspecified"/>
+    /// or <see cref="DateTimeKind.Local"/>, but <see cref="BigqueryParameterType.Timestamp"/> if its <c>Kind</c> is
+    /// <see cref="DateTimeKind.Utc"/>.
```

Comment:
This all seems unfortunately complex :( I suspect it will lead to confusion. Not sure what can be done about it though - unless DateTime *always* becomes DateTime, and possibly only Utc values are allowed?